src/ArrowFunctions.cpp

```diff
@@ -1078,31 +1084,88 @@ int cpp_writeMultiColToParquet(const char* filename, void* column_names,
           auto data_ptr = (uint8_t*)ptr_arr[i];
           parquet::ByteArrayWriter* ba_writer =
           static_cast<parquet::ByteArrayWriter*>(rg_writer->NextColumn());
-          int64_t count = 0;
-          int64_t byteIdx = 0;
+          if (objType_ptr[i] == SEGARRAY) {
+            int64_t byteIdx = 0;
 
-          // identify the starting byte index
-          if (x > 0){
-            byteIdx = idxQueue_str.front();
-            idxQueue_str.pop();
-          }
-          
-          while(count < batchSize) {
-            parquet::ByteArray value;
-            int16_t definition_level = 1;
-            value.ptr = reinterpret_cast<const uint8_t*>(&data_ptr[byteIdx]);
-            int64_t nextIdx = byteIdx;
-            while (data_ptr[nextIdx] != 0x00){
-              nextIdx++;
+            // identify the starting byte index
+            if (x > 0){
+              byteIdx = idxQueue_str.front();
+              idxQueue_str.pop();
+            }
+            
+            auto offset_ptr = (int64_t*)offset_arr[i];
+            int64_t offIdx = 0; // index into offsets
+
+            if (x > 0){
+              offIdx = idxQueue_segarray.front();
+              idxQueue_segarray.pop();
+            }
+            int64_t count = 0;
+            while (count < batchSize) { // ensures rowGroupSize maintained
+              int64_t segSize;
+              if (offIdx == numelems - 1) {
+                segSize = saSizes_ptr[i] - offset_ptr[offIdx];
+              }
+              else {
+                segSize = offset_ptr[offIdx+1] - offset_ptr[offIdx];
+              }
+              if (segSize > 0) {
+                for (int64_t s=0; s<segSize; s++) {
+                  int16_t def_lvl = 3;
+                  int16_t rep_lvl = (s == 0) ? 0 : 1;
+                  parquet::ByteArray value;
+                  value.ptr = reinterpret_cast<const uint8_t*>(&data_ptr[byteIdx]);
+                  int64_t nextIdx = byteIdx;
+                  while (data_ptr[nextIdx] != 0x00){
+                    nextIdx++;
+                  }
+                  value.len = nextIdx - byteIdx;
+                  ba_writer->WriteBatch(1, &def_lvl, &rep_lvl, &value);
+                  byteIdx = nextIdx + 1; // increment to start of next word
+                }
+              }
+              else {
+                // empty segment denoted by null value that is not repeated (first of segment) defined at the list level (1)
+                segSize = 1; // even though segment is length=0, write null to hold the empty segment
+                int16_t* def_lvl = new int16_t[segSize] { 1 };
+                int16_t* rep_lvl = new int16_t[segSize] { 0 };
+                ba_writer->WriteBatch(segSize, def_lvl, rep_lvl, nullptr);
+              }
+              offIdx++;
+              count++;
+            }
+            if (numLeft - count > 0) {
+              idxQueue_segarray.push(byteIdx);
+              idxQueue_segarray.push(offIdx);
```

Comment:
i noticed we are pushing `byteIdx` onto `idxQueue_segarray` here, but above we are popping it from `idxQueue_str`. I just wanted to double check that this is intentional

Reply:
Yes, `byteIdx` should be being pushed to`idxQueue_str`. Updated.